### ISSUE TITLE
can't return moduleType in on-resolve callback

### ISSUE THREAD

**Issue by ArrayZoneYour** [evanw/esbuild#2958]
Reproduce Demo

`test.js`
```js
import { __assign } from 'tslib'
console.log(__assign)
```

`build.js`
```js
require('esbuild').build({
  entryPoints: ['test.js'],
  outdir: "dist",
  platform: "neutral",
  bundle: true,
  format: 'cjs',
  plugins: [{
    name: 'resolver',
    setup(build) {
      build.onResolve({ filter: /.*/ }, args => {
        if (args.path === 'tslib') {
          return {
            path: path.resolve(require.resolve(args.path), '../modules/index.js'),
            // moduleType: 6 // ModuleESM_PackageJSON
          }
        }
      })
    }
  }]
})
```

`with plugin`

`dist/test.js`
```
// ...
var import_tslib = __toESM(require_tslib())
// ...
```

```
TypeError: Cannot destructure property '__extends' of 'import_tslib.default' as it is undefined
```

`without plugin`

`dist/test.js`
```
// ...
var import_tslib = __toESM(require_tslib(), 1)
// ...
```

```
[Function: assign]
```

**Comment by evanw** [comment]
I agree that it'd be nice if you could do this with a plugin, but this is not the right approach. The `onResolve` callback is called many times for the same module identifier so doing what you suggested is unsound. That would make it possible for plugins to return different things from different `onResolve` callbacks, which would be bad.

I'm changing this issue from a request for a very specific API to a request for the ability to do this (however it ends up being implemented).

**Comment by ArrayZoneYour** [comment]
As the issue is changed to a general API to handle module identifier, I want to show the other problems I use onResolve, onLoad callback.

1. the same module afterResolve can load different content
`src/index.ts`

```ts
import Bg from '@images/bg.png';
import Video from '@videos/video.mp4';
```

`./src/components/Banner.tsx`

```tsx
import Bg from '@images/bg.png'
import Video4 from '@videos/video4.mp4';

export { Bg, Video4 }
```

Currently, we can only set the same publicPath / assetName for assets. If developer use onResolve to customize their module resolution, the same module will have high possibility to load more times

```build.js
// Redirect all paths starting with "images/" to "./public/images/"
build.onResolve({ filter: /^@\// }, args => {
   return { path: customResolve(args) }
})```

build.onLoad({ filter: /\.png$/ }, async (args) => {
   let text = await fs.promises.readFile(args.path, 'utf8')
   return {
     contents: getPublicPath() + './images/' + text.split(/\s+/),
      loader: 'txt',
    }
})

build.onLoad({ filter: /\.mp4$/ }, async (args) => {
   let text = await fs.promises.readFile(args.path, 'utf8')
   await fs.writeFile(getOutputDir() +'/videos/' + text.split(/\s+/);
   return {
     contents: getPublicPath() + '/videos/' + text.split(/\s+/),
      loader: 'txt',
    }
})

2. Can custom "file" loader get benefits from DCE ?

If we use the same build.js from above, and the export `Video4`  from `./src/components/Banner.tsx` is not used in the other files, will the `writeFile` (onLoad operator for the module) skipped ?

**Comment by jwatzman** [comment]
I just came across this issue and would definitely love some API to either override the module metadata/type, or otherwise override the [default export semantics heuristics](https://esbuild.github.io/content-types/#default-interop). We have a large monorepo, with a single `package.json`, and multiple build targets for both client and server code. Trying to get all of the right semantics around our dependencies' default imports has been a nightmare. (The wisdom of having our repo set up like this, I'm not sure about, but it is what it is.) For various boring/annoying reasons, both actually setting `type: 'module'` and renaming a bunch of files `.mjs` aren't very palatable, but I'd love for a way to tell just esbuild to please build things (especially default imports) *as if* we had done that.

One approach I thought about was using an `onLoad` plugin to intercept the load of our main `package.json` and add `type: 'module'` into it before esbuild parsed it. That didn't work since esbuild doesn't seem to call `onLoad` for `package.json`. But I wonder if adding that behaviour might work as an API here -- either extending `onLoad` to do that, or adding a new plugin hook so that plugins could intercept and adjust the view of `package.json` that esbuild sees. It would allow setting `type: 'module'`, and potentially other package metadata? Not sure what all metadata you have in mind for this API and how much of it comes from `package.json` or elsewhere.

In any event, would love to see that or some other API in that direction. Thanks for all the work you do on esbuild!
